packages/react-dom/src/__tests__/ReactDOMComponent-test.js

```diff
@@ -146,6 +146,147 @@ describe('ReactDOMComponent', () => {
       }
     });
 
+    it('should warn for conflicting CSS shorthand updates', () => {
+      const container = document.createElement('div');
+      ReactDOM.render(
+        <div style={{font: 'foo', fontStyle: 'bar'}} />,
+        container,
+      );
+      expect(() =>
+        ReactDOM.render(<div style={{font: 'foo'}} />, container),
+      ).toWarnDev(
+        'Warning: Removing a style property during rerender (fontStyle) ' +
+          'when a conflicting property is set (font) can lead to styling ' +
+          "bugs. To avoid this, don't mix shorthand and non-shorthand " +
+          'properties for the same value; instead, replace the shorthand ' +
+          'with separate values.' +
+          '\n    in div (at **)',
+      );
+
+      // These updates are OK and don't warn:
+      ReactDOM.render(
+        <div style={{font: 'qux', fontStyle: 'bar'}} />,
+        container,
+      );
+      ReactDOM.render(
+        <div style={{font: 'foo', fontStyle: 'baz'}} />,
+        container,
+      );
+
+      expect(() =>
+        ReactDOM.render(
+          <div style={{font: 'qux', fontStyle: 'baz'}} />,
```

Comment:
Ugh, this is subtle. Had to read these tests a couple times. Nice to have a warning for it.